src/ast/passes/fold_literals.cpp

```diff
@@ -0,0 +1,450 @@
+{
+  if (auto *integer = dynamic_cast<Integer *>(op.expr)) {
+    if (op.op == Operator::BNOT) {
+      // Still positive.
+      return ast_.make_node<Integer>(~integer->value, Location(op.loc));
+    } else if (op.op == Operator::LNOT) {
+      // Still positive.
+      return ast_.make_node<Integer>(!integer->value, Location(op.loc));
+    } else if (op.op == Operator::MINUS) {
+      // Ensure that it is representable as a negative value.
+      if (integer->value >
+          static_cast<uint64_t>(std::numeric_limits<int64_t>::max()) + 1) {
+        op.addError() << "negative value will underflow";
+        return nullptr;
+      }
+      // Carefully make the conversion. We need to ensure that this does not
+      // overflow/underflow while doing the math.
+      if (integer->value <= 2) {
+        return ast_.make_node<NegativeInteger>(
+            -static_cast<int64_t>(integer->value), Location(op.loc));
+      } else {
+        int64_t value = -2;
+        value -= static_cast<int64_t>(integer->value - 2);
+        return ast_.make_node<NegativeInteger>(value, Location(op.loc));
+      }
+    }
+  } else if (auto *integer = dynamic_cast<NegativeInteger *>(op.expr)) {
+    if (op.op == Operator::BNOT) {
+      // Always positive.
+      return ast_.make_node<Integer>(static_cast<uint64_t>(~integer->value),
+                                     Location(op.loc));
+    } else if (op.op == Operator::LNOT) {
+      // Always positive.
+      return ast_.make_node<Integer>(static_cast<uint64_t>(!integer->value),
+                                     Location(op.loc));
+    } else if (op.op == Operator::MINUS) {
+      // Ensure that it doesn't overflow. We do this by ensuring that is
+      // representable as a positive number, casting, and then adding 1 to
+      // the unsigned form.
+      int64_t value = integer->value + 1;
+      value = -value;
+      return ast_.make_node<Integer>(static_cast<uint64_t>(value) + 1,
+                                     Location(op.loc));
+    }
+  }
+  return nullptr;
+}
+
+Expression *LiteralFolder::visit(Binop &op)
+{
+  visitAndReplace(&op.left);
+  visitAndReplace(&op.right);
+
+  // Check for string cases.
+  auto *str = dynamic_cast<String *>(op.left);
+  auto *other = op.right;
+  if (str == nullptr) {
+    str = dynamic_cast<String *>(op.right);
+    other = op.left;
+  }
+  if (str) {
+    // For whatever reason you are allowed to fold "foo"+3.
+    auto *integer = dynamic_cast<Integer *>(other);
+    if (op.op == Operator::PLUS && integer) {
+      if (integer->value >= static_cast<uint64_t>(str->value.size())) {
+        op.addWarning() << "literal string will always be empty";
+        return ast_.make_node<String>("", Location(op.loc));
+      }
+      return ast_.make_node<String>(str->value.substr(integer->value),
+                                    Location(op.loc));
+    }
+
+    // Check for another string.
+    auto *rs = dynamic_cast<String *>(other);
```

Comment:
Might be more cruft but since you checked above already you could have a variable indicating if you already know the other is not a string.

Reply:
I don't love extra state, so I'd rather leave as is. Easier for me to understand, at least personally.